tsdb/head_append.go

```diff
@@ -605,7 +610,12 @@ func (s *memSeries) appendHistogram(t int64, sh histogram.SparseHistogram, appen
 		return sampleInOrder, chunkCreated
 	}
 
-	if !chunkCreated {
+	if !chunkCreated && value.IsStaleNaN(sh.Sum) {
+		// Deliberately empty other fields so that any future histogram will result in a new chunk.
```

Comment:
Doesn't that mean that appending the stale marker itself will go into a new chunk?
Which would be wasteful, because at any regular end of a series, we would append a whole chunk with only the stale marker as the only content.

I think we either have to construct a histogram.SparseHistogram that is not seen as a counter reset (and ideally creates double-deltas of 0 to take the minimum amount of space), or we have to special-case the code within app.Appendable and app.AppendHistogram to deal with the stale-marker case appropriately.

Reply:
We could just use the last histogram with Sum set to StaleNaN then. And what special case do you have in mind? I did not exactly understand it.

Reply:
We could use the last histogram, with just the Sum changed, but that is unlikely to result in double-deltas of zero.

Because it's actually hard to create a histogram that will create double-deltas of zero, my idea was to special-case `app.Appendable` to always return "yup, it's appendable without problem" in case Sum is `StaleNaN`, and special-case `app.AppendHistogram` to just encode the Sum properly and set everything else to zero in case Sum is `StaleNaN`.

Reply:
But I believe if we get a new sample that is not stale, we do have to cut a new chunk at that time

Reply:
Yes, I think we should do that.

The idea is essentially to end the chunk with the stale marker.

But I think with the code as is in this PR, we would essentially create an own chunk thus for the stale marker.

Reply:
So perhaps we should have the following shortcut behavior in `app.Appendable` and `app.AppendHistogram`:

Method | Behavior if current sample is stale marker | Behavior if previous sample is stale marker
---|---|---
`app.Appendable` | Return `true` | return `false` (unless current sample is stale, too)
`app.AppendHistogram` | Append stale marker as `sum` and all other double-deltas and XOR's as zero | Append sample to a new chunk

Reply:
Yeah that is what I was trying to do, but facing some issues iterating it back. Will try it again tomorrow.